### PR TITLE
Implement chains support

### PR DESCRIPTION
**Pull request by e0ne**

### CODE REVIEW

File: chain.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package netlink
+
+import (
+	"fmt"
+)
+
+// ChainAttrs are the attributes of a Chain
+type Chain struct {
```
Comment (adrianchiris) [inline chain.go:8]: i wonder if its worth, similar to Filter[1] and Qdisc[2] to abstract this with an interface for consistency.

on the other hand, in contrast to Filter and Qdisc objects we dont have different types of  chains (currently...) so maybe its not worth it. 

just a thought :)

[1] https://github.com/vishvananda/netlink/blob/8e1ce9665a8f8ab9ed94d8830f08781c2951351c/filter.go#L8
[2] https://github.com/vishvananda/netlink/blob/8e1ce9665a8f8ab9ed94d8830f08781c2951351c/qdisc.go#L26

File: chain.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package netlink
+
+import (
+	"fmt"
+)
+
+// ChainAttrs are the attributes of a Chain
```
Comment (adrianchiris) [inline chain.go:7]: nit: ChainAttr -> Chain ?

File: chain_linux.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
+package netlink
+
+import (
+	"fmt"
+	"github.com/vishvananda/netlink/nl"
+	"golang.org/x/sys/unix"
+)
+
```
Comment (adrianchiris) [inline chain_linux.go:8]: would it make sense to add Unit test for this file ?

File: chain_linux.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	for _, m := range msgs {
+		msg := nl.DeserializeTcMsg(m)
+
+		attrs, err := nl.ParseRouteAttr(m[msg.Len():])
+		if err != nil {
+			return nil, err
+		}
+
+		// skip chains from other interfaces
+		if link != nil && msg.Ifindex != index {
```
Comment (adrianchiris) [inline chain_linux.go:98]: not a netlink expert :)  but i see you provide at L76 the ifIndex in the tc msg.
does the kernel returns chains that dont belong to the provided interface index ?

**Comment by adrianchiris** [comment]
@e0ne it would be great to add support for specifying chain in filter (for list/get/modify) as well as part of this PR or a separate one :)

see:
https://github.com/shemminger/iproute2/blob/9f366536edb5158343152604e82b968be46dbf26/tc/tc_filter.c#L189

File: chain_linux.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	msg := &nl.TcMsg{
+		Family:  nl.FAMILY_ALL,
+		Ifindex: index,
+		Parent:  chain.Parent,
+	}
+	req.AddData(msg)
+	req.AddData(nl.NewRtAttr(nl.TCA_CHAIN, nl.Uint16Attr(chain.Chain)))
+
+	_, err := req.Execute(unix.NETLINK_ROUTE, 0)
+	fmt.Println("6")
```
Comment (adrianchiris) [inline chain_linux.go:54]: nit: remove this

File: chain_linux.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		h.ensureIndex(base)
+		index = int32(base.Index)
+	}
+	msg := &nl.TcMsg{
+		Family:  nl.FAMILY_ALL,
+		Ifindex: index,
+		Parent:  chain.Parent,
+	}
+	req.AddData(msg)
+	req.AddData(nl.NewRtAttr(nl.TCA_CHAIN, nl.Uint16Attr(chain.Chain)))
```
Comment (adrianchiris) [inline chain_linux.go:51]: Chain attribute is 32 bits.

https://github.com/shemminger/iproute2/blob/49c63bc775d610c3dfd3db0d6501ac29c519967f/tc/tc_filter.c#L189

File: chain.go (hunk 1 above)
Comment (adrianchiris) [reply chain.go:8]: ack, makes sense to me.

File: chain_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+		t.Fatal(err)
+	}
+	if len(chains) != 1 {
+		t.Fatal("Failed to add chain")
+	}
+	if chains[0].Chain != *chainVal {
+		t.Fatalf("Incorrect chain added")
+	}
+	classId := MakeHandle(1, 1)
+	filter := &U32{
```
Comment (adrianchiris) [inline chain_test.go:74]: testing filter with chain attr should be part of filter_test.go IMO

File: chain_test.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		QdiscAttrs: QdiscAttrs{
+			LinkIndex: link.Attrs().Index,
+			Handle:    MakeHandle(0xffff, 0),
+			Parent:    HANDLE_INGRESS,
+		},
+	}
+	if err := QdiscAdd(qdisc); err != nil {
+		t.Fatal(err)
+	}
+	qdiscs, err := SafeQdiscList(link)
```
Comment (adrianchiris) [inline chain_test.go:45]: i think we can assume `QdiscAdd` works as expected (as its functionality is covered by Qdisc related tests.
so IMO you can remove L45-55 WDYT ?

File: cmd/ipset-test/main.go [hunk 9]
```diff
@@ -123,6 +125,22 @@ func cmdListAll(args []string) {
 	}
 }
 
+func cmdListFilter(args []string) {
```
Comment (adrianchiris) [inline cmd/ipset-test/main.go:128]: how is this related to ipset-test ?

File: filter_linux.go [hunk 10]
```diff
@@ -446,6 +450,10 @@ func (h *Handle) FilterList(link Link, parent uint32) ([]Filter, error) {
 				default:
 					detailed = true
 				}
+			case nl.TCA_CHAIN:
+				val := new(uint32)
+				*val = binary.LittleEndian.Uint32(attr.Value)
```
Comment (adrianchiris) [inline filter_linux.go:455]: system may be big-endian (e.g there are powerpc cpus that are big endian)

you can just use `netlink.native.Uint32()` like is being used in other places.

File: chain_test.go (hunk 8 above)
Comment (e0ne) [reply chain_test.go:45]: I prefer to keep these lines for consistency and to make TestChainAddDel more explicit in case of failure

File: chain_test.go (hunk 7 above)
Comment (e0ne) [reply chain_test.go:74]: done

File: chain_linux.go (hunk 3 above)
Comment (e0ne) [reply chain_linux.go:7]: tests added

File: chain_test.go (hunk 8 above)
Comment (e0ne) [reply chain_test.go:43]: even more: such tests helped me to find an issue with filter chain implementation

File: chain_linux.go (hunk 4 above)
Comment (adrianchiris) [reply chain_linux.go:96]: i see its done the same in QdiscList[1]
but not done in FilterList[2]

is it really needed? 
if there is no reason to keep this check, lets remove it.

[1] https://github.com/vishvananda/netlink/blob/e0988fd129c1858c092c73fd2c1bfd1a9d2af45d/qdisc_linux.go#L352
[2] https://github.com/vishvananda/netlink/blob/e0988fd129c1858c092c73fd2c1bfd1a9d2af45d/filter_linux.go#L363

File: chain_test.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+		t.Fatal(err)
+	}
+	link, err := LinkByName("foo")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := LinkSetUp(link); err != nil {
+		t.Fatal(err)
+	}
+	redir, err := LinkByName("bar")
```
Comment (adrianchiris) [inline chain_test.go:26]: why do you need "bar" link ?

File: chain_test.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+	chains, err := ChainList(link, HANDLE_INGRESS)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(chains) != 1 {
+		t.Fatal("Failed to add chain")
+	}
+	if chains[0].Chain != *chainVal {
+		t.Fatalf("Incorrect chain added")
+	}
```
Comment (adrianchiris) [inline chain_test.go:70]: can you also test for `chains[0].Parent` ?

File: filter_linux.go [hunk 13]
```diff
@@ -235,6 +223,9 @@ func (h *Handle) filterModify(filter Filter, flags int) error {
 		Info:    MakeHandle(base.Priority, nl.Swap16(base.Protocol)),
 	}
 	req.AddData(msg)
+	if filter.Attrs().Chain != nil {
+		req.AddData(nl.NewRtAttr(nl.TCA_CHAIN, nl.Uint32Attr(*filter.Attrs().Chain)))
+	}
 	req.AddData(nl.NewRtAttr(nl.TCA_KIND, nl.ZeroTerminated(filter.Type())))
```
Comment (adrianchiris) [inline filter_linux.go:230]: in case proto is `unix.RTM_DELTFILTER` i dont think we want to add all the below RTAttrs below

Comment (e0ne) [reply filter_linux.go:230]: In an original implementation in iproute2 there's no specific case for deletion: [1] and [2]


[1] https://github.com/shemminger/iproute2/blob/main/tc/tc_filter.c#L761
[2] https://github.com/shemminger/iproute2/blob/main/tc/tc_filter.c#L67-L86

Comment (adrianchiris) [reply filter_linux.go:230]: ack

**Review by adrianchiris — APPROVED** [review]
Thx for addressing my comments !

**Comment by aboch** [comment]
please squash your commits into one

**Comment by e0ne** [comment]
> please squash your commits into one

done. thanks for the feedback, @aboch
